src/Features/Core/Portable/Completion/CompletionService.cs

```diff
@@ -216,38 +220,38 @@ public virtual TextSpan GetDefaultCompletionListSpan(SourceText text, int caretP
             string filterText)
         {
             var helper = CompletionHelper.GetHelper(document);
-            return FilterItems(helper, items, filterText);
+            var itemsWithPatternMatch = new SegmentedList<(CompletionItem, PatternMatch?)>(items.Select(
+                item => (item, helper.GetMatch(item.FilterText, filterText, includeMatchSpans: false, CultureInfo.CurrentCulture))));
+
+            var builder = ImmutableArray.CreateBuilder<CompletionItem>();
+            FilterItems(helper, itemsWithPatternMatch, filterText, builder);
+            return builder.ToImmutable();
         }
 
-        internal virtual ImmutableArray<CompletionItem> FilterItems(
+        internal virtual void FilterItems(
            Document document,
-           ImmutableArray<(CompletionItem, PatternMatch?)> itemsWithPatternMatch,
-           string filterText)
+           IReadOnlyList<(CompletionItem, PatternMatch?)> itemsWithPatternMatch,
+           string filterText,
+           ImmutableArray<CompletionItem>.Builder builder)
         {
-            // Default implementation just drops the pattern matches and
-            // calls the public overload of FilterItems for compatibility.
-            return FilterItems(document, itemsWithPatternMatch.SelectAsArray(item => item.Item1), filterText);
+            // Default implementation just drops the pattern matches and builder, and
+            // calls the public overload of FilterItems instead for compatibility.
+            FilterItems(document, itemsWithPatternMatch.SelectAsArray(item => item.Item1), filterText);
         }
 
-        internal static ImmutableArray<CompletionItem> FilterItems(
-            CompletionHelper completionHelper,
-            ImmutableArray<CompletionItem> items,
-            string filterText)
-        {
-            var itemsWithPatternMatch = items.SelectAsArray(
-                item => (item, completionHelper.GetMatch(item.FilterText, filterText, includeMatchSpans: false, CultureInfo.CurrentCulture)));
-
-            return FilterItems(completionHelper, itemsWithPatternMatch, filterText);
-        }
+        // The FilterItems method might need to handle a large list of items when import completion is enabled and filter text is very short i.e. <= 1
+        // Therefore, use pooled list to avoid repeated (potentially LOH) allocations.
+        private static readonly ObjectPool<List<(CompletionItem item, PatternMatch? match)>> s_listOfItemMatchPairPool = new(factory: () => new(), size: 2);
```

Comment:
why a list, and not an arraybuilder?  (generally speaking, this is feedback on consistency across this part of hte codebase).

Reply:
OK, changed it back to List, should be all consistent now